### PR TITLE
Add scalar stats estimation for Cast (v2)

### PR DESCRIPTION
**Pull request by losipiuk**
Replaces https://github.com/Teradata/presto/pull/573

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/cost/SymbolStatsAssertion.java [hunk 1]
```diff
@@ -80,4 +80,16 @@ public SymbolStatsAssertion distinctValuesCountUnknown()
         assertTrue(isNaN(statistics.getDistinctValuesCount()), "expected unknown distinctValuesCount but got " + statistics.getDistinctValuesCount());
         return this;
     }
+
+    public SymbolStatsAssertion dataSize(double expected)
+    {
+        assertEquals(statistics.getDataSize(), expected, "dataSize mismatch");
```
Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/cost/SymbolStatsAssertion.java]: shouldn't it use @kokosing comparison?

File: presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsCalculator.java [hunk 2]
```diff
@@ -101,5 +105,54 @@ protected SymbolStatsEstimate visitLiteral(Literal node, Void context)
...
+            SymbolStatsEstimate sourceStats = process(node.getExpression());
+            TypeSignature targetType = TypeSignature.parseTypeSignature(node.getType());
+
+            // todo - make this general postprocessing rule.
+            double distinctValuesCount = sourceStats.getDistinctValuesCount();
+            double lowValue = sourceStats.getLowValue();
+            double highValue = sourceStats.getHighValue();
+
+            if (isIntegralType(targetType)) {
+                if (isFinite(lowValue)) {
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsCalculator.java]: I think round behaves correctly with `-Inf` and `Inf` double values. I think it's just enough to test for `NaN`

File: presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsCalculator.java [hunk 3]
```diff
@@ -101,5 +105,54 @@ protected SymbolStatsEstimate visitLiteral(Literal node, Void context)
...
+                    lowValue = Math.round(lowValue);
+                }
+                if (isFinite(highValue)) {
+                    highValue = Math.round(highValue);
+                }
+                if (isFinite(lowValue) && isFinite(highValue)) {
+                    double integersInRange = highValue - lowValue + 1;
+                    if (isNaN(distinctValuesCount) || distinctValuesCount > integersInRange) {
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsCalculator.java]: shouldn't you preserve `NaN` distinct values count? I think it should be:
```if (!isNaN(distinctValuesCount) && distinctValuesCount > integersInRange) {```

File: presto-main/src/test/java/com/facebook/presto/cost/TestScalarStatsCalculator.java [hunk 4]
```diff
@@ -72,6 +77,87 @@ public void testLiteral()
                 .nullsFraction(1.0);
     }
 
+    @Test
+    public void testCastDoubleToBigint()
+    {
+        Map<Symbol, Type> types = ImmutableMap.of(new Symbol("a"), DOUBLE);
+
+        PlanNodeStatsEstimate inputStatistics = PlanNodeStatsEstimate.builder()
```
Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/cost/TestScalarStatsCalculator.java:85]: really I would also rename `PlanNodeStatsEstimate` to `RelationStatsEstimate`

File: presto-main/src/test/java/com/facebook/presto/cost/TestScalarStatsCalculator.java [hunk 5]
```diff
@@ -72,6 +77,87 @@ public void testLiteral()
...
+    public void testCastDoubleToShortRange()
+    {
+        Map<Symbol, Type> types = ImmutableMap.of(new Symbol("a"), DOUBLE);
+
+        PlanNodeStatsEstimate inputStatistics = PlanNodeStatsEstimate.builder()
+                .addSymbolStatistics(new Symbol("a"), SymbolStatsEstimate.builder()
+                        .setNullsFraction(0.3)
+                        .setLowValue(1.6)
+                        .setHighValue(3.3)
+                        .setDistinctValuesCount(10)
```
Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/cost/TestScalarStatsCalculator.java:113]: also maybe refactor `DistinctValuesCount` to just `NDV`

File: presto-main/src/test/java/com/facebook/presto/cost/TestScalarStatsCalculator.java [hunk 6]
```diff
@@ -72,6 +77,87 @@ public void testLiteral()
...
+        assertCalculate(new Cast(new SymbolReference("a"), "bigint"), inputStatistics, types)
+                .lowValue(2.0)
+                .highValue(3.0)
+                .distinctValuesCount(2)
+                .nullsFraction(0.3)
+                .dataSizeUnknown();
+    }
+
+    @Test
+    public void testCastBigintToDouble()
```
Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/cost/TestScalarStatsCalculator.java:149]: add check that `NaN` NDV is preserved

File: presto-main/src/test/java/com/facebook/presto/cost/TestScalarStatsCalculator.java (hunk 4 above)
Comment (losipiuk) [reply presto-main/src/test/java/com/facebook/presto/cost/TestScalarStatsCalculator.java:85]: Not at this point.

File: presto-main/src/test/java/com/facebook/presto/cost/TestScalarStatsCalculator.java (hunk 5 above)
Comment (losipiuk) [reply presto-main/src/test/java/com/facebook/presto/cost/TestScalarStatsCalculator.java:113]: maybe. Not now though.

File: presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsCalculator.java (hunk 2 above)
Comment (losipiuk) [reply presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsCalculator.java]: I decided to leave it as this. It is not super important.
I will add todo to handle trimming domain so we make range narrower due to cast.

**Comment by losipiuk** [comment]
Merged

**Pull request closed without merging**